roadmap/implementers-guide/src/runtime/scheduler.md

```diff
@@ -1,3 +1,227 @@
+transaction pool, as other extrinsics.
+
+There should be no reason to place bids long before we want to claim a slot. If
+spam protection mechanism are effective you should just send the bid, when you
+want to have a core in the near future ... bids will have some lifetime
+specified. If they don't manage to win a slot within that lifetime, they are
+dropped. The user would then need to bid again.
+
+Needs:
+
+1. Expose: Realistic price to pay right now. (E.g. expose recent max/lowest
+   prices/average).
+2. Actual utilization - related to price, but nevertheless good to know
+   explicitly. If there are lots of free cores right now, it is "safe" to go with
+   the minimum bid.
+3. Some form of lifetime ... Some time frame the user can expect to either get a
+   core or know for sure he won't get any. Needs to be exposed somehow? - HOW?
+4. The user does not need to be able to specify a start ... once he submits a
+   bid, we treat this as "get me a core asap". No need to be able to schedule in
+   advance.
+
+I think it is fine for the lifetime to be implicit (determined by the system,
+e.g. by a 10 block auction mechanism), but we need to expose what happened to
+the bid. The actual auction mechanism used, should not be exposed.
+
+#### Node Side Implementation
+
+On the node side validators should be able to do some pre-processing on bids,
+for handling spam. In particular this is only relevant, if there are lots of
+bids. Ideally on the node side we would already:
+
+1. Check signatures, remove invalid bids and disconnect peers providing such
+   bids.
+2. If a certain threshold of pending bids is reached, we can start dropping
+   lowest bids.
+
+TODO: How does this work with transactions? Can we even check signatures on the
+node side? If we don't check signatures on the node side, then how can we be
+spam resistant?
+
+Then the block producer should put bids into the block, limited by some maximum
+amount (don't overflow runtime memory). Higher bids take precedence obviously.
+
+If exposed how many parathread cores are available, that information can be
+taken into account by the block producer: E.g. if there are 3 cores available,
+10 blocks would mean approximately 30 bids are useful to have in the runtime
+right now. It would not be useful to put 10000 bids into a block for example.
+Block producers should then likely expire bids in the transaction pool already,
+if they don't make it after some to be specified lifetime.
+
+Expiring bids makes sense to keep resource usage down, but also helps with
+keeping expectations: If we say you bid now for a block asap, it would be out of
+expectations if your bid finally gets accepted two days later. This aspect would
+not need to be implemented on the node side though, but is already an
+optimization. We could also just import everything into the runtime and let the
+runtime drop what is too low (if resource limits allow for it).
```

Comment:
Do you mean that even though bids won, the block producer could decide to reject them? Would the block producer then talk to the auction pallet to re-add them to the pool with highest priority? And fixed spots? I.e. they can't be outbid in the next auction even though their bids are < min_bid_in_full_pool?

Reply:
I would go for something very simple for filtering bids. If we assume that we can only bid for blocks later than height N, we can filter by the price, relevance(lifetime bids too far into the future can be kept in the txpool), and just weight. The parathread can monitor the extrinsic and know when it  makes it to a block, and then wait for cores to be assigned during the selected lifetime.

When there is a lot of demand, the bids might not even make it through in a block, and this is totally visibile as with any extrinsic.

Reply:
If you do end-up reusing the normal transaction pool for this, this might be relevant: https://paritytech.github.io/substrate/master/sp_runtime/transaction_validity/struct.ValidTransaction.html#structfield.longevity